File: utilities/src/main/kotlin/lunalauf/rms/utilities/network/server/ClientHandler.java
package lunalauf.rms.utilities.network.server;

import javafx.application.Platform;
import javafx.beans.Observable;
import javafx.collections.FXCollections;
import javafx.collections.ListChangeListener;
import javafx.collections.ObservableList;
import javafx.concurrent.Task;
import lunalauf.rms.utilities.network.util.ConnectionInitiationHelper;

import java.net.SocketException;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.function.Consumer;

public class ClientHandler {

    private final ExecutorService threadPool;
    private final LunaLaufAPI api;
    private final ObservableList<Client> clients;

    public ClientHandler(ExecutorService threadPool, LunaLaufAPI api) {
        this.threadPool = threadPool;
        this.api = api;
        this.clients = FXCollections.observableList(new ArrayList<>(),
                client -> new Observable[]{client.statusProperty()});
    }

    public void handleClient(Client client) {
        CommunicationInitiationTask task = new CommunicationInitiationTask(client);
        threadPool.execute(task);
    }

    private class CommunicationInitiationTask extends Task<Void> {

        private static final int comTestTimeout = 2000; // ms
        private final Client client;

        private CommunicationInitiationTask(Client client) {
            this.client = client;
        }

        @Override
        protected Void call() throws Exception {
            if (initialCommunicationTest(client)) {
                Platform.runLater(() -> ClientHandler.this.registerClient(client));
                client.startListening(threadPool, api);
            } else {
                client.close();
            }
            return null;
        }

        private boolean initialCommunicationTest(Client client) {
            try {
                client.setTimeout(comTestTimeout);
                String synMessage = client.receive();

                String ackMessage = ConnectionInitiationHelper.getAckMessage(synMessage);
                client.send(ackMessage);

                String synAckMessage = client.receive();
                String expectedSynAckMessage = ConnectionInitiationHelper.getAckMessage(ackMessage);

                return expectedSynAckMessage.equals(synAckMessage);
            } catch (Exception e) {
                return false;
            } finally {
                try {
                    client.resetTimeout();
                } catch (SocketException ignored) {
                }
            }
        }
    }

    private void registerClient(Client client) {
        Client existing = null;
        for (Client otherClient : clients) {
            if (otherClient.getRemoteAddress().equals(client.getRemoteAddress()))
                if (otherClient.getStatus() == -1)
                    existing = otherClient;
        }
        if (existing != null)
            clients.set(clients.indexOf(existing), client);
        else
            clients.add(client);
    }

    public ObservableList<Client> getClients() {
        return clients;
    }

    public void setOnLostClients(Consumer<List<? extends Client>> listener) {
        this.clients.addListener((ListChangeListener<Client>) change -> {
            if (change.next() && change.wasUpdated()) {
                var lostClients = change.getList().stream().filter(c -> c.getStatus() < 0).toList();
                if (!lostClients.isEmpty())
                    listener.accept(lostClients);
            }
        });
    }

}

File: utilities/src/main/kotlin/lunalauf/rms/utilities/network/server/ClientHandler.kt
package lunalauf.rms.utilities.network.server

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import lunalauf.rms.modelapi.ModelState
import lunalauf.rms.utilities.network.util.ConnectionInitiationHelper
import org.slf4j.LoggerFactory
import java.net.SocketException

class ClientHandler(
    private val modelState: StateFlow<ModelState>
) {
    companion object {
        private const val TIMEOUT_COMMUNICATION_TEST = 2000 // ms
    }

    private val logger = LoggerFactory.getLogger(this::class.java)

    private val scope = CoroutineScope(Dispatchers.IO)
    private val _clients = MutableStateFlow<List<Client>>(listOf())
    val clients get() = _clients.asStateFlow()

    private var onLostClient: (Client) -> Unit = {}

    fun handleClient(client: Client) {
        scope.launch {
            if (initialCommunicationTest(client)) {
                registerClient(client)
                client.startListening(modelState)
            } else {
                client.close()
            }
        }
    }

    private fun initialCommunicationTest(client: Client): Boolean {
        return try {
            client.setTimeout(TIMEOUT_COMMUNICATION_TEST)
            val synMessage = client.receive() ?: return false
            val ackMessage = ConnectionInitiationHelper.getAckMessage(synMessage)
            client.send(ackMessage)
            val synAckMessage = client.receive()
            val expectedSynAckMessage: String = ConnectionInitiationHelper.getAckMessage(ackMessage)
            val result = expectedSynAckMessage == synAckMessage
            logger.info("Communication test {}: {}", if (result) "successful" else "failed", client.remoteAddress)
            result
        } catch (e: Exception) {
            logger.warn("Communication test failed due to an exception: {}", client.remoteAddress, e)
            false
        } finally {
            try {
                client.resetTimeout()
            } catch (e: SocketException) {
                logger.error("Failed to reset socket timeout. Behaviour unpredictable: {}", client.remoteAddress, e)
            }
        }
    }

    private fun registerClient(client: Client) {
        client.onLost { onLostClient(client) }

        var existing: Client? = null
        _clients.value.forEach {
            if (it.remoteAddress == client.remoteAddress)
                if (it.status.value == -1)
                    existing = it
        }
        _clients.update {
            it.toMutableList().apply {
                if (existing != null) set(it.indexOf(existing), client)
                else add(client)
            }
        }
        logger.info("Client registered: {}", client.remoteAddress)
    }

    fun onLostClient(action: (Client) -> Unit) {
        onLostClient = action
    }

    fun removeClient(client: Client) {
        if (client.status.value < 0)
            _clients.update {
                it.toMutableList().apply {
                    remove(client)
                }
            }
    }
}
